File: palindromelinklist.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
#O(N for both)
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        #1)find mid
        slow=head
        fast=head
        while fast and fast.next:
            fast=fast.next.next
            slow=slow.next
        #2)reverse 2nd half
        def reverse(self,head):
            cur=head
            prev=None
            while cur:
                temp=cur.next
                cur.next=prev
                prev=cur
                cur=temp
            return (prev)
        revhalf=reverse(self,slow)
        #3)traverse both halfs simultaneously till reversed list ends and if values mismatch>false
        while revhalf:
            if revhalf.val!=head.val:
                return False
            revhalf=revhalf.next
            head=head.next
        return True

File: PalindromeLinkList.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
 //TC: O(n)
 //SC: O(1)
 class Solution {
    public boolean isPalindrome(ListNode head) {

        if(head == null || head.next == null) return true;

        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
       
        fast = reverse(slow.next);
        slow.next = null;
        ListNode curr = head;
        
        //compare

        while(fast != null){

            if(curr.val != fast.val) return false;
            curr = curr.next;
            fast = fast.next;
        }

        return true;
    }

    public ListNode reverse(ListNode head){
        if(head == null || head.next == null) return head;

        ListNode prev = null;
        ListNode curr = head;
        System.out.println(curr.val);
        while(curr != null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
          System.out.println(prev.val);
        return prev;
    }
}

 